io/plugins/eu.esdihumboldt.hale.io.gml/src/eu/esdihumboldt/hale/io/gml/reader/internal/wfs/WfsBackedGmlInstanceCollection.java

```diff
@@ -0,0 +1,472 @@
+		// XXX Use WFSVersion
+		switch (version) {
+		case "1.1.0":
+			return "MAXFEATURES";
+		case "2.0.0":
+		case "2.0.2":
+			return "COUNT";
+		default:
+			throw new IllegalArgumentException("Unsupported WFS version");
+		}
+	}
+
+	/**
+	 * Iterates over {@link Instance}s in the GML stream retrieved from the WFS.
+	 * 
+	 * The iterator will not load more features per WFS GetFeature request than
+	 * specified in the {@link WfsBackedGmlInstanceCollection}. If the WFS query
+	 * yields more results, multiple GetFeature request will be issued
+	 * transparently to the WFS until all results have been retrieved or the
+	 * maximum number of features as specified in the
+	 * {@link WfsBackedGmlInstanceCollection} was reached.
+	 * 
+	 * @author Florian Esser
+	 */
+	public class WfsBackedGmlInstanceIterator implements InstanceIterator {
+
+		private GmlInstanceCollection currentCollection;
+		private GmlInstanceIterator iterator;
+		private int totalFeaturesProcessed;
+
+		/**
+		 * Create the iterator
+		 */
+		public WfsBackedGmlInstanceIterator() {
+			createNextIterator();
+		}
+
+		/**
+		 * Closes the iterator of the currently active GmlInstanceCollection and
+		 * creates a new one for the next WFS request. If the new request yields
+		 * no result, this {@link WfsBackedGmlInstanceIterator} is closed.
+		 */
+		private void proceedOrClose() {
+			iterator.close();
+			createNextIterator();
+
+			if (!iterator.hasNext()) {
+				close();
+			}
+		}
+
+		private void createNextIterator() {
+			URI nextUri;
+			try {
+				nextUri = calculateNextUri();
+			} catch (URISyntaxException e) {
+				throw new IllegalStateException(e.getMessage(), e);
+			}
+
+			log.debug(MessageFormat.format("Creating new iterator for URL \"{0}\"",
+					nextUri.toString()));
+
+			currentCollection = new GmlInstanceCollection(new DefaultInputSupplier(nextUri),
+					sourceSchema, restrictToFeatures, ignoreRoot, strict, ignoreNamespaces,
+					crsProvider, ioProvider);
+			iterator = currentCollection.iterator();
+		}
+
+		private URI calculateNextUri() throws URISyntaxException {
+			URIBuilder builder = new URIBuilder(baseUri);
+
+			// Use STARTINDEX value in primordial URI as offset
+			int offset = 0;
+			if (primordialQueryParams.containsKey("STARTINDEX")) {
+				try {
+					offset = Integer.parseInt(primordialQueryParams.get("STARTINDEX"));
+				} catch (NumberFormatException e) {
+					// Ignore if invalid
+				}
+			}
+
+			// STARTINDEX is 0-based
+			builder.addParameter("STARTINDEX", Integer.toString(offset + totalFeaturesProcessed));
+			builder.addParameter(getMaxFeaturesParameterName(wfsVersion),
+					Integer.toString(WfsBackedGmlInstanceCollection.this.featuresPerRequest));
+
+			return builder.build();
+		}
+
+		/**
+		 * @see java.util.Iterator#hasNext()
+		 */
+		@Override
+		public boolean hasNext() {
+			if (isClosed()) {
+				return false;
+			}
+
+			if (maxNumberOfFeatures != UNLIMITED && totalFeaturesProcessed >= maxNumberOfFeatures) {
```

Comment:
I did not see the place where we stop issuing new requests when there are no more features though we still did not hit the max number of features. Can you please point me to it?